src/monitor/StakerMonitor.ts

```diff
@@ -0,0 +1,288 @@
+import { ethers } from 'ethers';
+import { IDatabase } from '@/database';
+import { EventProcessor } from './EventProcessor';
+import { ConsoleLogger, Logger } from './logging';
+import {
+  MonitorConfig,
+  MonitorStatus,
+  StakeDepositedEvent,
+  StakeWithdrawnEvent,
+  DelegateeAlteredEvent,
+} from './types';
+import { STAKER_ABI } from './constants';
+
+export class StakerMonitor {
+  private readonly db: IDatabase;
+  private readonly provider: ethers.Provider;
+  private readonly contract: ethers.Contract;
+  private readonly logger: Logger;
+  private readonly eventProcessor: EventProcessor;
+  private readonly config: MonitorConfig;
+  private isRunning: boolean;
+  private processingPromise?: Promise<void>;
+  private lastProcessedBlock: number;
+
+  constructor(config: MonitorConfig) {
+    this.config = config;
+    this.db = config.database;
+    this.provider = config.provider;
+    this.contract = new ethers.Contract(
+      config.stakerAddress,
+      STAKER_ABI,
+      config.provider,
+    );
+    this.logger = new ConsoleLogger(config.logLevel);
+    this.eventProcessor = new EventProcessor(this.db, this.logger);
+    this.isRunning = false;
+    this.lastProcessedBlock = config.startBlock;
+  }
+
+  async start(): Promise<void> {
+    if (this.isRunning) {
+      this.logger.warn('Monitor is already running');
+      return;
+    }
+
+    this.isRunning = true;
+    this.logger.info('Starting Staker Monitor', {
+      network: this.config.networkName,
+      chainId: this.config.chainId,
+      address: this.config.stakerAddress,
+    });
+
+    // Load last checkpoint if exists
+    const checkpoint = await this.db.getCheckpoint('staker-monitor');
+    if (checkpoint) {
+      this.lastProcessedBlock = checkpoint.last_block_number;
+      this.logger.info('Loaded checkpoint', {
+        blockNumber: this.lastProcessedBlock,
+      });
+    }
+
+    this.processingPromise = this.processLoop();
+  }
+
+  async stop(): Promise<void> {
+    if (!this.isRunning) {
+      return;
+    }
+
+    this.isRunning = false;
+    if (this.processingPromise) {
+      await this.processingPromise;
+    }
+    this.logger.info('Staker Monitor stopped');
+  }
+
+  private async processLoop(): Promise<void> {
+    while (this.isRunning) {
```

Comment:
Looks good! Can you confirm I'm reading it right:

It seems to me this will keep processing blocks at a clip of `maxBlockRange` until it is caught up to the latest "safe" block based on the `confirmations` config param. After that it sleeps for the `pollInterval` and wakes up to see if there are more blocks.

Reply:
this is right